Mathlib/Analysis/Meromorphic/Divisor/Basic.lean

```diff
@@ -120,138 +120,68 @@ ordered group, where addition, comparison, min and max of divisors are defined
+  add D₁ D₂ := mk_of_mem (D₁ + D₂) <| add_mem D₁.memAddSubgroup D₂.memAddSubgroup
 
-/-- Divisors have a negative -/
 instance : Neg (DivisorOn U) where
-  neg D := {
-    toFun := -D
-    supportWithinDomain' := by
-      intro x hx
-      rw [Function.support_neg', Function.mem_support, ne_eq] at hx
-      exact D.supportWithinDomain hx
-    supportDiscreteWithinDomain' := D.supportDiscreteWithinDomain.neg
-  }
-
-/-- Helper lemma for the `simp` tactic: the function of the negative divisor
-is the negative of the associated function. -/
-@[simp]
-lemma coe_neg {D : DivisorOn U} : (↑(-D) : 𝕜 → ℤ) = -(D: 𝕜 → ℤ) := rfl
+  neg D := mk_of_mem (-D) <| neg_mem D.memAddSubgroup
+
+instance : Sub (DivisorOn U) where
+  sub D₁ D₂ := mk_of_mem (D₁ - D₂) <| sub_mem D₁.memAddSubgroup D₂.memAddSubgroup
 
-/-- Divisors have scalar multiplication with natural numbers -/
 instance : SMul ℕ (DivisorOn U) where
-  smul n D := {
-    toFun := fun z ↦ n * D z
-    supportWithinDomain' := by
-      intro x hx
-      simp at hx
-      exact D.supportWithinDomain hx.2
-    supportDiscreteWithinDomain' := by
-      filter_upwards [D.supportDiscreteWithinDomain]
-      intro x hx
-      simp [hx]
-  }
-
-/-- Helper lemma for the `simp` tactic: the function of a scalar product
-(natural number)·divisor is the scalar product of the natural number with the
-associated function of the divisor. -/
-@[simp]
-lemma coe_nsmul {D : DivisorOn U} {n : ℕ} : (↑(n • D) : 𝕜 → ℤ) = n • (D : 𝕜 → ℤ) := rfl
+  smul n D := mk_of_mem (n • D) <| nsmul_mem D.memAddSubgroup n
 
-/-- Divisors have scalar multiplication with integers -/
 instance : SMul ℤ (DivisorOn U) where
-  smul n D := {
-    toFun := fun z ↦ n * D z
-    supportWithinDomain' := by
-      intro x hx
-      simp at hx
-      exact D.supportWithinDomain hx.2
-    supportDiscreteWithinDomain' := by
-      filter_upwards [D.supportDiscreteWithinDomain]
-      intro _ hx
-      simp [hx]
-  }
-
-/-- Helper lemma for the `simp` tactic: the function of a scalar product
-(integer)·divisor is the scalar product of the integer with the associated
-function of the divisor. -/
-@[simp]
-lemma coe_zsmul {D : DivisorOn U} {n : ℤ} : (↑(n • D) : 𝕜 → ℤ) = n • (D : 𝕜 → ℤ) := rfl
+  smul n D := mk_of_mem (n • D) <| zsmul_mem D.memAddSubgroup n
+
+@[simp] lemma coe_zero : ((0 : DivisorOn U) : 𝕜 → ℤ) = 0 := rfl
+@[simp] lemma coe_add (D₁ D₂ : DivisorOn U) : (↑(D₁ + D₂) : 𝕜 → ℤ) = D₁ + D₂ := rfl
+@[simp] lemma coe_neg (D : DivisorOn U) : (↑(-D) : 𝕜 → ℤ) = -(D : 𝕜 → ℤ) := rfl
+@[simp] lemma coe_sub (D₁ D₂ : DivisorOn U) : (↑(D₁ - D₂) : 𝕜 → ℤ) = D₁ - D₂ := rfl
+@[simp] lemma coe_nsmul (D : DivisorOn U) (n : ℕ) : (↑(n • D) : 𝕜 → ℤ) = n • (D : 𝕜 → ℤ) := rfl
+@[simp] lemma coe_zsmul (D : DivisorOn U) (n : ℤ) : (↑(n • D) : 𝕜 → ℤ) = n • (D : 𝕜 → ℤ) := rfl
+
+instance : AddCommGroup (DivisorOn U) :=
+  Function.Injective.addCommGroup (M₁ := DivisorOn U) (M₂ := 𝕜 → ℤ)
+    _ coe_injective coe_zero coe_add coe_neg coe_sub coe_nsmul coe_zsmul
```

Comment:
We transport all the proofs about the axioms for the group structure from the group structure on functions.

Reply:
Roger.